model_dev/entity_extraction/PURE/PURE_data_setup.py

```diff
@@ -0,0 +1,211 @@
+
+                        tokenized_child_word = tokenizer.tokenize(child_word)
+                        if len(tokenized_child_word) == 1:
+                            tmp.append(int(sentences[0].index(tokenized_child_word[0])) + 1)
+                            tmp.append(int(sentences[0].index(tokenized_child_word[0])) + 1)
+                            if [int(sentences[0].index(tokenized_child_word[0])) + 1, int(sentences[0].index(tokenized_child_word[0])) + 1, "BASE"] not in ner:
+                                ner.append([int(sentences[0].index(tokenized_child_word[0])) + 1, int(sentences[0].index(tokenized_child_word[0])) + 1, "BASE"])
+                        else:
+                            tmp.append(int(sentences[0].index(tokenized_child_word[0])) + 1)
+                            tmp.append(int(sentences[0].index(tokenized_child_word[-1])) + 1)
+                            if [int(sentences[0].index(tokenized_child_word[0])) + 1, int(sentences[0].index(tokenized_child_word[-1])) + 1, "BASE"] not in ner:
+                                ner.append([int(sentences[0].index(tokenized_child_word[0])) + 1, int(sentences[0].index(tokenized_child_word[-1])) + 1, "BASE"])
+                        tmp.append(relation["label"])
+                        rel.append(tmp)
+
+            final_sent['doc_key'] = str(tmp_index)
+            final_sent['sentences'] = sentences
+            final_sent['ner'] = [ner]
+            final_sent['predicted_ner'] = [ner]
+            final_sent['relations'] = [rel]
+            final_sent['right_to_left'] = right_to_left
+            final_sent['left_to_right'] = left_to_right
+
+            #print(final_sent)
+            #print("===")
+
+        if final_sent:
+            final_sent_arr.append(final_sent)
+            final_sent = {}
+            tmp_index += 1
+    
+"""
+train, test = train_test_split(final_sent_arr, test_size=0.1)
+val, test = train_test_split(test, test_size=0.5)
+"""
+"""
+    Scikit-learn's train_test_split will shuffle the data before creating the tran-test sets, which will result in inconsistent number of right_to_left and left_to_right
+    relationships in test, train, and dev sets. To fix that, first the array with all the sentences will be sorted by right_to_left counter
+    for the respective sentence, and then an index after which no sentence has right_to_let relationship is found. We have to make sure that majority of right_to_left relationships
+    are in train set. In this case 50% sentences containing 
+"""
+
+total_sent = len(final_sent_arr)
+
+# sorting the array with right_to_left count
+sorted_by_direction = sorted(final_sent_arr, key=lambda d: d['right_to_left'], reverse=True) 
+
+# getting the index after which no sentence will have a right_to_left relationship
+right_to_left_boundary = get_right_to_left_boundary(sorted_by_direction)
```

Comment:
I am not sure about this part. Would using the stratify parameter of sklearn's split solve your problem instead of doing all this?

Reply:
Also, for the entity training model (with which we got 0.75 strict f1 for "base"), this is the file PURE_ent_data_setup.py which uses train_test_split. Thanks!

Reply:
You should try the stratify parameter, it will ensure everything is the same between train, test and val.

Reply:
(+ @himesh257 )

Reply:
one thing that confuses me is the usage of stratify where we don’t have a "y", you see? @CVxTz do you know how it would work in this case?

Reply:
And yes, PURE will first predict the entities and then relationships. For simplicity and speed, I am populating the entities so I can skip that step and directly run the relationship model for now. So all in all, it will eventually have the same split

Reply:
@himesh257 You can create an indicator about the most common relationship direction in the sentence and use this as the stratify value.
Can you share the resulting files from your different splits somewhere ?
Thanks

Reply:
Realized that stratify isn’t needed anymore, updated scripts